### PR TITLE
[BUG] in `CNNClassifier` and `CNNRegressor`, fix hidden layer activation function parameter passing

### LINKED ISSUE #8878: [BUG] `CNNRegressor` uses the same activation function (linear by default) for all layers
# Motivation

`CNNRegressor` is currently an adaptation of the CNN classifier from [this implementation](https://github.com/hfawaz/dl-4-tsc/blob/master/classifiers/cnn.py).

The architecture in `sktime` consists of several convolutional + pooling layers, followed by a dense layer with a single output node. At present, the same activation function is applied after each convolutional layer **and** at the output.

For regression tasks:
- A linear activation at the output layer is typically a valid choice, as it allows unrestricted continuous predictions.
- However, using a linear activation in the convolutional layers is problematic, since it prevents the network from learning non-linear feature representations.

The cited paper introduces the CNN for classification and uses `sigmoid` for all layers (including output, which is valid for classification)

# Proposed Changes

- Split the current activation argument into two:
  - `activation_hidden`: activation function for convolutional layers
  - `activation_output`: activation function for the final dense layer

- Default values in CNNRegressor:
  - `activation_hidden="sigmoid"` (to follow the cited paper, though `relu` is generally more common)
  - `activation_output="linear"`

- Deprecate the existing `activation` argument in `CNNRegressor` for backward compatibility.

If this sounds good, I’d be happy to open a PR.

### PR DESCRIPTION
**Pull request by noxthot**
#### Reference Issues/PRs
Example: Fixes #8878. 

#### What does this implement/fix? Explain your changes.
Does not falsly override activation function for hidden layers in CNNRegressor and corrects documentation.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
-

#### Did you add any tests for the change?
No

#### Any other comments?
-

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [-] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [-] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [-] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [-] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/regression/deep_learning/cnn.py [hunk 1]
```diff
@@ -129,7 +129,6 @@ def __init__(
             avg_pool_size=self.avg_pool_size,
             n_conv_layers=self.n_conv_layers,
             filter_sizes=self.filter_sizes,
-            activation=self.activation,
```
Comment (yarnabrina) [inline sktime/regression/deep_learning/cnn.py:132]: Allowing user to specify a value and then completely ignoring that specification is not right in my opinion. The above docstring mismatch should be fixed, and that change we should merge. But this one is a breaking change and the current state of silently overriding user specification should not happen as is.

Comment (noxthot) [reply sktime/regression/deep_learning/cnn.py:132]: This commit is basically reverting (one) bug that was introduced with https://github.com/sktime/sktime/commit/4e2f2c29eaca0b9ed808cc582a996a910519e543

The specification is not ignored, but only used for the output layer, just like the doc string states.

This is a quick fix to restore the original behavior.

More information can be found in the linked issue https://github.com/sktime/sktime/issues/8878

Comment (yarnabrina) [reply sktime/regression/deep_learning/cnn.py:132]: I've seen your posts before my review. I am not much involved in regression module, and won't review/comment without further knowledge, but your current revert modifies the estimator behaviour as it was expected by its users (I don't know how much it's used) since 2023. It is a breaking change, not with respect to change in signature but change in behaviour, even if it's for the better (which I do not know and am not commenting on).

My suggestion will be to split this PR: one to fix the docstring which is straightforwarsd and should be addressed. And second is change wrt the behaviour where @achieveordie and other active core dev of regression module can review/discuss how to handle this.

Comment (noxthot) [reply sktime/regression/deep_learning/cnn.py:132]: Sorry for the long text; without much context, the removal of that line really looks a bit odd.

So to add more context for one of the regression module core devs:

The activation function for the hidden layers is chosen to get non-linearity into the system (to be able to learn complex patterns) and to improve/ensure convergence.

The activation function for the output layer is primarily chosen to allow the neural network to output values that are used for classification and regression.

So for the hidden layers you would use a non-linear function and "non-linear" is crucial here.

For the output layer, you typically use `sigmoid` or `softmax` (or something similar) for classification, and `linear` or `relu` for regression (depending on the range of the function you would like to approximate), since you typically would not want an activation "distorting" the output.

I am quite sure it just accidentially happened with [this pull request](https://github.com/sktime/sktime/pull/4075), that the same activation function was suddenly not only applied to the output layer (like the doc states and stated), but also to the hidden layers. This change was the breaking change:
- It broke application of `CNNRegressor` with default settings (since `linear` makes no sense for hidden layers)
- It broke every application of  `CNNRegressor` where you need a `linear` output activation (and this was the standard case for me so far) since you can't have it at the same time with non-linearity in the hidden layers.
- It changed the behaviour for everyone who set `activation` for `CNNClassificier` and `CNNRegressor` and updated to the version that contained this pull request, since suddenly, this `activation` became also applied to the hidden layers (before that `sigmoid` was used in both cases for the hidden layers).

As far as I can see, this was neither the intention of the [related issue](https://github.com/sktime/sktime/issues/4067), nor documented in the [release notes of Version 0.15.1 (2023-01-12)](https://www.sktime.net/en/latest/changelog.html#id839).

I think, the best would be to deprecate `activation` and introduce `activation_hidden` and `activation_output`. This increases flexibility and also makes it possible to "simulate" any behaviour we had in the past:
- The original behaviour (< 0.15.1) would just be `activation_hidden=sigmoid` and `activation_output` set to `activation`.
- The behaviour since 2023 (although faulty; >= 0.15.1) would be to set `activation_hidden` and `activation_output` to `activation`.

Since either users with a version < 0.15.1 are affected, or users >= 0.15.1, one needs to decide which of the above two options should be the default setting when those users have `activation` currently set.

And also, one needs to decide what should happen when `activation` was not set by those users:
- For the `CNNClassifier` it would simply stay the same (already using `sigmoid`).
- For the `CNNRegressor` we would need to move away from the `linear` activation in the hidden layers. -> "Breaking change" for the user; but it is necessary since having `linear` in the hidden layer really is not a good idea if you want to train a performant model.

**Comment by achieveordie** [comment]
The concerns raised by @noxthot are definitely valid. Currently, `CNNNetwork` is applying the same activation function set by the user and that might negatively affect its performance. This is indeed a legitimate bug and requires a deeper discussion.

There are at least two ways to go about this:
1. Follow @noxthot's suggestion, deprecate `activation` in favor of `activation_hidden` and `activation_output`.
2. Looking at the general behavior of other models in this space like [`MCDNN`](https://github.com/sktime/sktime/blob/d587dc372e4d74ef5d18b2c41a31b9a6576bbac8/sktime/networks/mcdcnn.py#L95) and [`MLP`](https://github.com/sktime/sktime/blob/d587dc372e4d74ef5d18b2c41a31b9a6576bbac8/sktime/networks/mlp.py#L65), I see that the `activation` parameter is generally only applied to the output layer.

I also haven't looked into the all the models so it might also be worthwhile to see whether all the models exhibit the second behavior, it it does, we should correct the behavior in `CNNNetwork`. There's also a legitimate case where the user wants to specifically modify the activation functions in the hidden layers as well, in which case we'd have to rework the module to accommodate for such customization.

I'm a little rusty with things here since I haven't been active in this space, and I'd like the active core developers to chime in with their opinion. If there's no one spearheading the development of the DL module, I'd be happy to step up and take a second look at it.

**Comment by noxthot** [comment]
Thanks for your response, @achieveordie .

I have checked all classes that are derived from `BaseDeepNetwork` and made a table to show which one expose the `activation` parameter (in case they do, the parameter would be used to control the `activation` of the hidden layers):

Class | activation exposed
-- | --
CNNNetwork | ✓
CNTCNetwork |  (should be there according to the documentation)
FCNNetwork |  
InceptionTimeNetwork |  
LSTMFCNNetwork |  
MACNNNetwork |  
MCDCNNNetwork |  
MLPNetwork |  
ResNetNetwork |  
RNNNetwork |  
TapNetNetwork |  

And those classes are used by the following models to prepare the hidden layers of the neural network. The output layers are usually implemented directly in the model classes. Here is a list of all the models indicating whether `activation` is exposed and for what it is used.

Model (used by) | Activation exposed | Hidden | Output
-- | -- | -- | --
CNNClassifier | ✓ | ✓ | ✓
CNNRegressor | ✓ | ✓ | ✓
CNTCRegressor |   |   |  
FCNRegressor | ✓ |   | ✓
InceptionTimeClassifier |   |   |  
InceptionTimeRegressor |   |   |  
LSTMFCNClassifier |   |   |  
LSTMFCNRegressor |   |   |  
MACNNClassifier | ✓ |   | ✓
MACNNRegressor | ✓ (not documented) |  | 
MCDCNNClassifier | ✓ |   | ✓
MCDCNNRegressor | ✓ |   | ✓
MLPClassifier | ✓ |   | ✓
MLPRegressor | ✓ |   | ✓
ResNetClassifier | ✓ |   | ✓
ResNetRegressor | ✓ |   | ✓
SimpleRNNClassifier | ✓  |   | ✓
SimpleRNNRegressor | ✓ |   | ✓
TapNetClassifier | ✓ |   | ✓
TapNetRegressor | ✓ |   | ✓ (init mismatch between doc and code)

This list also makes it evident that `activation` should not be applied to `CNNClassifier` and `CNNRegressor` for constistency reasons.

Still I think it would be nice to have the ability of configuring the hidden layer activations as an option.

I'd propose the following four-stage process:
- Pull Request 1: Restore the original behavior of `CNNClassifier` and `CNNRegressor` by not using the `activation` for the hidden layers. Additionally introduce `activation_hidden` and `activation_output` for these two models, allowing for proper deprecation of `activation` and according warning to the users on how to simulate < 0.15.1 or >= 0.15.1 behaviour (this is only possible when the two new options exist).

- Pull Request 2: Fix the documentation for `TapNetRegressor`, `CNTCNetwork` and remove `activation` from `MACNNRegressor` (since it is neither used nor documented).

- Pull Request 3: Offer `activation_hidden` and `activation_output` for all the models which do not expose `activation` currently. -> new feature; no compatibility issues.

- Pull Request 4: Offer `activation_hidden` and `activation_output` for all the models which already expose `activation` and deprecate `activation`. -> no compatibility issue, since `activation_output` can safely just use `activation` when present and be set to the same default as `activation`; but users will get a warning to use `activation_output` in future.

Again, I'd be happy to work on this issue once the core devs agree on the procedure.

**Comment by yarnabrina** [comment]
@noxthot your steps make sense. If none objects, I think there's usually a 3-7 day period we assume, you can proceed with this. I think @achieveordie can also help in reviews.

If you want, you can join our Discord where they may be more active than on Github.

**Review by fkiraly — DISMISSED** [review]
I would agree with the fix - I would also say that deprecation is not necessary. Even though behaviour changes, it is buggy behaviour that gets fixed (bugs are excepted from the need to deprecate)

`sktime` discord: https://discord.com/invite/54ACzaFsn7

**Comment by fkiraly** [comment]
> I'm a little rusty with things here since I haven't been active in this space, and I'd like the active core developers to chime in with their opinion. If there's no one spearheading the development of the DL module, I'd be happy to step up and take a second look at it.

Long time no speak, @achieveordie! The efforts in the DL space are focused on `pytorch-forecasting`, and we want to move the `sktime` DL modules to `torch` as well, based on learnings from `pytorch-forecasting`. There is no one leading this in `sktime`, so your participation would be more than welcome!
Some devs focusing currently on DL and `torch` APIs are @agobbifbk, @phoeenniixx, @PranavBhatP.

> The concerns raised by @noxthot are definitely valid. Currently, `CNNNetwork` is applying the same activation function set by the user and that might negatively affect its performance. This is indeed a legitimate bug and requires a deeper discussion.
> 
> There are at least two ways to go about this:
> 
>     1. Follow @noxthot's suggestion, deprecate `activation` in favor of `activation_hidden` and `activation_output`.
> 
>     2. Looking at the general behavior of other models in this space like [`MCDNN`](https://github.com/sktime/sktime/blob/d587dc372e4d74ef5d18b2c41a31b9a6576bbac8/sktime/networks/mcdcnn.py#L95) and [`MLP`](https://github.com/sktime/sktime/blob/d587dc372e4d74ef5d18b2c41a31b9a6576bbac8/sktime/networks/mlp.py#L65), I see that the `activation` parameter is generally only applied to the output layer.
> 
> 
> I also haven't looked into the all the models so it might also be worthwhile to see whether all the models exhibit the second behavior, it it does, we should correct the behavior in `CNNNetwork`. There's also a legitimate case where the user wants to specifically modify the activation functions in the hidden layers as well, in which case we'd have to rework the module to accommodate for such customization.

My suggestion would be to use `activation` for the output layer if that would make the parameterizations more consistent, and introduce separate parameter names for the hidden layers. That would also avoid the need for deprecation.

**Comment by noxthot** [comment]
Thanks for your responses.

I have implemented the following changes:
- `CNNRegression`:
  - `activation` is only applied to the output layer now (previously output layer and hidden layer) and defaults to `linear` (like it used to)
  - `activation_hidden` was added and defaults to `sigmoid` (original behaviour prior `0.15.1`)
  
  - `CNNClassifier`:
  - `activation` is only applied to the output layer now (previously output layer and hidden layer) and defaults to `softmax` (like it used to)
  - `activation_hidden` was added and defaults to `sigmoid`

So this should be ready to be reviewed, restore the behavior as of versions prior `0.15.1`, fix the bug in https://github.com/sktime/sktime/issues/8878 and also extend flexibility in configuring those two models.

Since deprecation is not needed according to @fkiraly, I have only added two seperate issues:
https://github.com/sktime/sktime/issues/8889
https://github.com/sktime/sktime/issues/8890

**Review by fkiraly — APPROVED** [review]
Thanks.

While we aer at it, could you kindly change the docstring parameter formatting to be consistent with the rest of the repo? E:g., `kernsel_size : int, default = 7` (only single space before the `:`

**Comment by fkiraly** [comment]
made some minor changes:

* removed test skips
* moved author credit for @noxthot to the estimator tag (author credits are now managed in tags)

**Comment by noxthot** [comment]
Thanks, @fkiraly.
 
> While we aer at it, could you kindly change the docstring parameter formatting to be consistent with the rest of the repo? E:g., `kernsel_size : int, default = 7` (only single space before the `:`

Done.

**Review by achieveordie — APPROVED** [review]
LGTM too, thanks!

**Comment by achieveordie** [comment]
> Long time no speak, @achieveordie! The efforts in the DL space are focused on pytorch-forecasting, and we want to move the sktime DL modules to torch as well, based on learnings from pytorch-forecasting. There is no one leading this in sktime, so your participation would be more than welcome!

Hello there, @fkiraly :) It’s great to see how well things are going! I had been meaning to find an entry point but I was stuck in analysis paralysis. Thanks for pointing me in the appropriate direction. I'll take a few days going over PRs and discussions independently before joining the conversations on Discord.

**Comment by noxthot** [comment]
I investigated the failed test and think it was a temporary error when accessing [timeseriesclassification.com](https://timeseriesclassification.com):
`URLDownloader failed: Could not download/extract dataset from https://timeseriesclassification.com/aeon-toolkit/Beef.zip: HTTP Error 403: Forbidden`

It only failed for one of the jobs, it worked locally and the URL is valid. Could you initiate a re-run?

**Comment by noxthot** [comment]
> I investigated the failed test and think it was a temporary error when accessing [timeseriesclassification.com](https://timeseriesclassification.com): `URLDownloader failed: Could not download/extract dataset from https://timeseriesclassification.com/aeon-toolkit/Beef.zip: HTTP Error 403: Forbidden`
> 
> It only failed for one of the jobs, it worked locally and the URL is valid. Could you initiate a re-run?

The same test succeeded in https://github.com/sktime/sktime/actions/runs/18257549295/job/51981902355?pr=8917
So I am confident it was a temporary failure.

**Comment by fkiraly** [comment]
@noxthot, this analysis here
https://github.com/sktime/sktime/pull/8880#issuecomment-3365453018

Is very useful! Can you move it to a new "todo" issue about parametrization of deep learning methods if such an issue does not exist yet? Including the table.

Because once we merge this PR it will be less visible to people working on the deep learning module.

**Comment by noxthot** [comment]
> @noxthot, this analysis here [#8880 (comment)](https://github.com/sktime/sktime/pull/8880#issuecomment-3365453018)
> 
> Is very useful! Can you move it to a new "todo" issue about parametrization of deep learning methods if such an issue does not exist yet? Including the table.
> 
> Because once we merge this PR it will be less visible to people working on the deep learning module.

already created an issue two days ago, but added the table now -> https://github.com/sktime/sktime/issues/8889
